File: apps/mobile/POS ADBD/Library/Managers/PurchasesManager.swift
//
//  PurchasesManager.swift
//  POS ADBD
//
//  RevenueCat wrapper for MERCHANT subscription renewals (6m / 12m via StoreKit).
//  App User ID = merchant_{publicId} for webhook mapping to Subscription.
//

import Foundation
import RevenueCat

struct RenewPackageInfo {
    let productId: String
    let title: String
    let priceLabel: String
    /// Store price in currency major units.
    let priceAmount: Double
}

enum PurchasesManager {
    static let entitlementMerchant = "merchant_subscription"
    /// Must match App Store Connect + RevenueCat product identifiers.
    static let productSemiAnnual = "anyrent_merchant_semi_annual"
    static let productAnnual = "anyrent_merchant_annual"

    private static var configured = false
    /// Cached packages keyed by store product id (from last offerings fetch).
    private static var packageByProductId: [String: Package] = [:]

    static var isConfigured: Bool { configured }

    static func configure() {
        let apiKey = Bundle.main.object(forInfoDictionaryKey: "REVENUECAT_API_KEY") as? String
        let trimmed = apiKey?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !trimmed.isEmpty, !trimmed.hasPrefix("$(") else {
            NSLog("[PurchasesManager] REVENUECAT_API_KEY missing — Purchases not configured")
            return
        }
        guard !configured else { return }
        #if DEBUG
        Purchases.logLevel = .debug
        #else
        Purchases.logLevel = .info
        #endif
        Purchases.configure(withAPIKey: trimmed)
        configured = true
        NSLog("[PurchasesManager] RevenueCat configured")
        syncFromCurrentUser()
    }

    /// Call after login / cold start when User.account() exists.
    static func syncFromCurrentUser() {
        guard configured else { return }
        guard let user = User.account(),
              user.role == .merchant,
              let merchantId = user.merchantId,
              merchantId > 0 else {
            logOut()
            return
        }
        logInMerchant(merchantId: merchantId)
    }

    static func logInMerchant(merchantId: Int) {
        guard configured else { return }
        let appUserId = "merchant_\(merchantId)"
        Purchases.shared.logIn(appUserId) { customerInfo, created, error in
            if let error {
                NSLog("[PurchasesManager] logIn failed: \(error.localizedDescription)")
                return
            }
            let active = customerInfo?.entitlements.active.keys.joined(separator: ",") ?? ""
            NSLog("[PurchasesManager] logIn ok user=\(appUserId) created=\(created ?? false) active=\(active)")
        }
    }

    static func logOut() {
        guard configured else { return }
        guard !Purchases.shared.isAnonymous else { return }
        Purchases.shared.logOut { _, error in
            if let error {
                NSLog("[PurchasesManager] logOut failed: \(error.localizedDescription)")
            } else {
                NSLog("[PurchasesManager] logOut ok")
            }
        }
    }

    static func hasMerchantEntitlement(_ info: CustomerInfo) -> Bool {
        info.entitlements[entitlementMerchant]?.isActive == true
    }

    static func getRenewPackages(completion: @escaping ([RenewPackageInfo]?, Error?) -> Void) {
        guard configured else {
            completion(nil, NSError.errorWithOwnMessage(
                message: "Purchases not configured".localized(),
                domain: "PurchasesManager"
            ))
            return
        }
        Purchases.shared.getOfferings { offerings, error in
            if let error {
                NSLog("[PurchasesManager] getOfferings failed: \(error.localizedDescription)")
                completion(nil, error)
                return
            }
            var byId: [String: Package] = [:]
            if let current = offerings?.current {
                for pkg in current.availablePackages {
                    byId[pkg.storeProduct.productIdentifier] = pkg
                }
            }
            if let all = offerings?.all {
                for offering in all.values {
                    for pkg in offering.availablePackages {
                        if byId[pkg.storeProduct.productIdentifier] == nil {
                            byId[pkg.storeProduct.productIdentifier] = pkg
                        }
                    }
                }
            }
            packageByProductId = byId
            let ordered = [productSemiAnnual, productAnnual]
            let renew: [RenewPackageInfo] = ordered.compactMap { id in
                guard let pkg = byId[id] else { return nil }
                return RenewPackageInfo(
                    productId: id,
                    title: titleForProduct(id),
                    priceLabel: pkg.storeProduct.localizedPriceString,
                    priceAmount: (pkg.storeProduct.price as NSDecimalNumber).doubleValue
                )
            }
            completion(renew, nil)
        }
    }

    static func purchase(
        productId: String,
        completion: @escaping (CustomerInfo?, Error?, Bool) -> Void
    ) {
        guard configured else {
            completion(nil, NSError.errorWithOwnMessage(
                message: "Purchases not configured".localized(),
                domain: "PurchasesManager"
            ), false)
            return
        }
        guard let pkg = packageByProductId[productId] else {
            completion(nil, NSError.errorWithOwnMessage(
                message: "Package not found — refresh offerings first".localized(),
                domain: "PurchasesManager"
            ), false)
            return
        }
        Purchases.shared.purchase(package: pkg) { _, customerInfo, error, userCancelled in
            if userCancelled {
                completion(nil, nil, true)
                return
            }
            if let error {
                NSLog("[PurchasesManager] purchase failed: \(error.localizedDescription)")
                completion(nil, error, false)
                return
            }
            NSLog("[PurchasesManager] purchase ok active=\(customerInfo?.entitlements.active.keys.joined(separator: ",") ?? "")")
            completion(customerInfo, nil, false)
        }
    }

    static func restore(completion: @escaping (CustomerInfo?, Error?) -> Void) {
        guard configured else {
            completion(nil, NSError.errorWithOwnMessage(
                message: "Purchases not configured".localized(),
                domain: "PurchasesManager"
            ))
            return
        }
        Purchases.shared.restorePurchases { customerInfo, error in
            if let error {
                NSLog("[PurchasesManager] restore failed: \(error.localizedDescription)")
                completion(nil, error)
                return
            }
            NSLog("[PurchasesManager] restore ok")
            completion(customerInfo, nil)
        }
    }

    private static func titleForProduct(_ productId: String) -> String {
        switch productId {
        case productSemiAnnual: return "6 months".localized()
        case productAnnual: return "12 months".localized()
        default: return productId
        }
    }
}

File: apps/mobile-android/app/src/main/java/com/anyrent/pos/billing/PurchasesManager.kt
package com.anyrent.pos.billing

import android.app.Activity
import android.app.Application
import android.util.Log
import com.anyrent.pos.BuildConfig
import com.anyrent.pos.data.SessionStore
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.LogLevel
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.PurchasesError
import com.revenuecat.purchases.getOfferingsWith
import com.revenuecat.purchases.logInWith
import com.revenuecat.purchases.logOutWith
import com.revenuecat.purchases.purchaseWith
import com.revenuecat.purchases.restorePurchasesWith
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine

data class RenewPackage(
    val productId: String,
    val title: String,
    val priceLabel: String,
    /** Store price in currency major units (from amountMicros). */
    val priceAmount: Double,
)

/**
 * RevenueCat wrapper for MERCHANT subscription renewals (6m / 12m via Store IAP).
 *
 * App User ID = `merchant_{publicId}` so webhooks can map to our Subscription row.
 * Non-merchant roles never log in to Purchases.
 */
object PurchasesManager {
    private const val TAG = "PurchasesManager"
    const val ENTITLEMENT_MERCHANT = "merchant_subscription"
    /** Play / App Store product ids — must match RevenueCat + store consoles. */
    const val PRODUCT_SEMI_ANNUAL = "anyrent_merchant_semi_annual"
    const val PRODUCT_ANNUAL = "anyrent_merchant_annual"

    @Volatile
    private var configured = false

    /** Cached packages keyed by store product id (from last offerings fetch). */
    @Volatile
    private var packageByProductId: Map<String, Package> = emptyMap()

    fun isConfigured(): Boolean = configured

    fun configure(app: Application) {
        val apiKey = BuildConfig.REVENUECAT_API_KEY.trim()
        if (apiKey.isEmpty()) {
            Log.w(TAG, "REVENUECAT_ANDROID_API_KEY empty — Purchases not configured")
            return
        }
        if (configured) return
        Purchases.logLevel = if (BuildConfig.DEBUG) LogLevel.DEBUG else LogLevel.INFO
        Purchases.configure(
            PurchasesConfiguration.Builder(app, apiKey).build(),
        )
        configured = true
        Log.i(TAG, "RevenueCat configured")
        syncFromSession()
    }

    /** Call after login / on cold start when session already exists. */
    fun syncFromSession() {
        if (!configured) return
        val role = SessionStore.role
        val merchantId = SessionStore.merchantId
        if (role == "MERCHANT" && merchantId != null && merchantId > 0) {
            logInMerchant(merchantId)
        } else {
            logOut()
        }
    }

    fun logInMerchant(merchantId: Int) {
        if (!configured) return
        val appUserId = "merchant_$merchantId"
        Purchases.sharedInstance.logInWith(
            appUserId,
            onError = { error -> Log.e(TAG, "logIn failed: ${error.message}") },
            onSuccess = { customerInfo, created ->
                Log.i(TAG, "logIn ok user=$appUserId created=$created active=${customerInfo.entitlements.active.keys}")
            },
        )
    }

    fun logOut() {
        if (!configured) return
        if (!Purchases.sharedInstance.isAnonymous) {
            Purchases.sharedInstance.logOutWith(
                onError = { error -> Log.e(TAG, "logOut failed: ${error.message}") },
                onSuccess = { Log.i(TAG, "logOut ok") },
            )
        }
    }

    fun hasMerchantEntitlement(info: CustomerInfo): Boolean =
        info.entitlements[ENTITLEMENT_MERCHANT]?.isActive == true

    suspend fun getRenewPackages(): Result<List<RenewPackage>> =
        suspendCancellableCoroutine { cont ->
            if (!configured) {
                cont.resume(Result.failure(IllegalStateException("Purchases not configured")))
                return@suspendCancellableCoroutine
            }
            Purchases.sharedInstance.getOfferingsWith(
                onError = { error ->
                    Log.e(TAG, "getOfferings failed: ${error.message}")
                    cont.resume(Result.failure(PurchasesException(error)))
                },
                onSuccess = { offerings ->
                    val allPackages = buildList {
                        offerings.current?.availablePackages?.let { addAll(it) }
                        offerings.all.values.forEach { offering ->
                            addAll(offering.availablePackages)
                        }
                    }
                    val byId = LinkedHashMap<String, Package>()
                    for (pkg in allPackages) {
                        byId.putIfAbsent(pkg.product.id, pkg)
                    }
                    packageByProductId = byId
                    val orderedIds = listOf(PRODUCT_SEMI_ANNUAL, PRODUCT_ANNUAL)
                    val renew = orderedIds.mapNotNull { id ->
                        val pkg = byId[id] ?: return@mapNotNull null
                        RenewPackage(
                            productId = id,
                            title = titleForProduct(id),
                            priceLabel = pkg.product.price.formatted,
                            priceAmount = pkg.product.price.amountMicros / 1_000_000.0,
                        )
                    }
                    cont.resume(Result.success(renew))
                },
            )
        }

    suspend fun purchase(activity: Activity, productId: String): Result<CustomerInfo> =
        suspendCancellableCoroutine { cont ->
            if (!configured) {
                cont.resume(Result.failure(IllegalStateException("Purchases not configured")))
                return@suspendCancellableCoroutine
            }
            val pkg = packageByProductId[productId]
            if (pkg == null) {
                cont.resume(
                    Result.failure(
                        IllegalStateException("Package not found for $productId — refresh offerings first"),
                    ),
                )
                return@suspendCancellableCoroutine
            }
            val params = PurchaseParams.Builder(activity, pkg).build()
            Purchases.sharedInstance.purchaseWith(
                params,
                onError = { error, userCancelled ->
                    if (userCancelled) {
                        cont.resume(Result.failure(PurchaseCancelledException()))
                    } else {
                        Log.e(TAG, "purchase failed: ${error.message}")
                        cont.resume(Result.failure(PurchasesException(error)))
                    }
                },
                onSuccess = { _, customerInfo ->
                    Log.i(TAG, "purchase ok active=${customerInfo.entitlements.active.keys}")
                    cont.resume(Result.success(customerInfo))
                },
            )
        }

    suspend fun restore(): Result<CustomerInfo> =
        suspendCancellableCoroutine { cont ->
            if (!configured) {
                cont.resume(Result.failure(IllegalStateException("Purchases not configured")))
                return@suspendCancellableCoroutine
            }
            Purchases.sharedInstance.restorePurchasesWith(
                onError = { error ->
                    Log.e(TAG, "restore failed: ${error.message}")
                    cont.resume(Result.failure(PurchasesException(error)))
                },
                onSuccess = { customerInfo ->
                    Log.i(TAG, "restore ok active=${customerInfo.entitlements.active.keys}")
                    cont.resume(Result.success(customerInfo))
                },
            )
        }

    private fun titleForProduct(productId: String): String = when (productId) {
        PRODUCT_SEMI_ANNUAL -> "6 months"
        PRODUCT_ANNUAL -> "12 months"
        else -> productId
    }
}

class PurchaseCancelledException : Exception("Purchase cancelled")

class PurchasesException(val purchasesError: PurchasesError) :
    Exception(purchasesError.message)
